rasa/core/policies/rule_policy.py

```diff
@@ -524,27 +595,78 @@ def _run_prediction_on_trackers(
+                    running_tracker, domain, interpreter, gold_action_name
                 )
-                if error_message:
-                    error_messages.append(error_message)
+                if collect_sources:
+                    self._collect_sources(
+                        running_tracker,
+                        predicted_action_name,
+                        gold_action_name,
+                        prediction_source,
+                    )
+                else:
+                    # to be able to remove only rules turns from the dialogue history
+                    # for ML policies,
+                    # we need to know which rules were used in ML trackers
+                    if (
+                        not tracker.is_rule_tracker
+                        and predicted_action_name == gold_action_name
+                    ):
+                        rules_used_in_stories.add(prediction_source)
+
+                    error_messages += self._check_prediction(
+                        running_tracker,
+                        predicted_action_name,
+                        gold_action_name,
+                        prediction_source,
+                    )
 
                 running_tracker.update(event)
 
-        return error_messages
+        return error_messages, rules_used_in_stories
 
-    def _find_contradicting_rules(
+    def _collect_rule_sources(
         self,
         rule_trackers: List[TrackerWithCachedStates],
-        all_trackers: List[TrackerWithCachedStates],
         domain: Domain,
         interpreter: NaturalLanguageInterpreter,
     ) -> None:
+        self._run_prediction_on_trackers(
+            rule_trackers, domain, interpreter, collect_sources=True
+        )
+
+    def _find_contradicting_and_used_in_stories_rules(
+        self,
+        trackers: List[TrackerWithCachedStates],
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+    ) -> Tuple[List[Text], Set[Text]]:
+        return self._run_prediction_on_trackers(
+            trackers, domain, interpreter, collect_sources=False
+        )
+
+    def _analyze_rules(
+        self,
+        rule_trackers: List[TrackerWithCachedStates],
+        all_trackers: List[TrackerWithCachedStates],
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+    ) -> List[Text]:
+        """Analyze learned rules by running prediction on training trackers.
+
+        This method collects error messages for contradicting rules
+        and creates the lookup for rules that are not present in the stories.
+
+        Args:
+            rule_trackers: The list of the rule trackers.
+            all_trackers: The list of all trackers.
+            domain: The domain.
+            interpreter: Interpreter which can be used by the polices for featurization.
+
+        Returns:
+             The list of rules that are not present in the stories.
```

Comment:
let's avoid duplicating the type in the docstring
```suggestion
             Rules that are not present in the stories.
```